### ISSUE TITLE
building gcc for c++11 threads

### ISSUE THREAD

**Issue by aquiles2k** [mxe/mxe#768]
I'm building for c++11 threads using the two stage compilation explained here: http://stackoverflow.com/a/24538951

To be able to use googles breakpad i only added "--disable-sjlj-exceptions" and "--with-dwarf2" to the configure step on the second stage. 

This worked until gcc 4.9.2 but since 5.1 the compilation fails on the second stage with:
/Users/tommy/mxe/tmp-gcc-i686-w64-mingw32.shared/gcc-5.1.0/libgcc/libgcc2.c:2256: undefined reference to `__EH_FRAME_BEGIN__'

Snip of last log lines:
/bin/sh /Users/tommy/mxe/tmp-gcc-i686-w64-mingw32.shared/gcc-5.1.0/libgcc/../mkinstalldirs .
ln -s -f libgcc.map libgcc.map.def && if [ ! -d ./shlib ]; then mkdir ./shlib; else true; fi && /Users/tommy/mxe/tmp-gcc-i686-w64-mingw32.shared/gcc-5.1.0.build/./gcc/xgcc -B/Users/tommy/mxe/tmp-gcc-i686-w64-mingw32.shared/gcc-5.1.0.build/./gcc/ -L/Users/tommy/mxe/usr/i686-w64-mingw32.shared/lib -L/Users/tommy/mxe/usr/mingw/lib -isystem /Users/tommy/mxe/usr/i686-w64-mingw32.shared/include -isystem /Users/tommy/mxe/usr/mingw/include -B/Users/tommy/mxe/usr/i686-w64-mingw32.shared/bin/ -B/Users/tommy/mxe/usr/i686-w64-mingw32.shared/lib/ -isystem /Users/tommy/mxe/usr/i686-w64-mingw32.shared/include -isystem /Users/tommy/mxe/usr/i686-w64-mingw32.shared/sys-include    -O2 -I/Users/tommy/mxe/tmp-gcc-i686-w64-mingw32.shared/gcc-5.1.0/libgcc/../winsup/w32api/include -g -O2 -DIN_GCC  -DCROSS_DIRECTORY_STRUCTURE  -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wstrict-prototypes -Wmissing-prototypes -Wold-style-definition  -isystem ./include   -g -DIN_LIBGCC2 -fbuilding-libgcc -fno-stack-protector -Dinhibit_libc -pthread -shared -nodefaultlibs libgcc.map.def -Wl,--out-implib,./shlib/libgcc_s.a.tmp -o ./shlib/libgcc_s_dw2-1.dll.tmp -g -O2 -B./ _chkstk_s.o _chkstk_ms_s.o _muldi3_s.o _negdi2_s.o _lshrdi3_s.o _ashldi3_s.o _ashrdi3_s.o _cmpdi2_s.o _ucmpdi2_s.o _clear_cache_s.o _trampoline_s.o __main_s.o _absvsi2_s.o _absvdi2_s.o _addvsi3_s.o _addvdi3_s.o _subvsi3_s.o _subvdi3_s.o _mulvsi3_s.o _mulvdi3_s.o _negvsi2_s.o _negvdi2_s.o _ctors_s.o _ffssi2_s.o _ffsdi2_s.o _clz_s.o _clzsi2_s.o _clzdi2_s.o _ctzsi2_s.o _ctzdi2_s.o _popcount_tab_s.o _popcountsi2_s.o _popcountdi2_s.o _paritysi2_s.o _paritydi2_s.o _powisf2_s.o _powidf2_s.o _powixf2_s.o _powitf2_s.o _mulsc3_s.o _muldc3_s.o _mulxc3_s.o _multc3_s.o _divsc3_s.o _divdc3_s.o _divxc3_s.o _divtc3_s.o _bswapsi2_s.o _bswapdi2_s.o _clrsbsi2_s.o _clrsbdi2_s.o _fixunssfsi_s.o _fixunsdfsi_s.o _fixunsxfsi_s.o _fixsfdi_s.o _fixdfdi_s.o _fixxfdi_s.o _fixunssfdi_s.o _fixunsdfdi_s.o _fixunsxfdi_s.o _floatdisf_s.o _floatdidf_s.o _floatdixf_s.o _floatundisf_s.o _floatundidf_s.o _floatundixf_s.o _divdi3_s.o _moddi3_s.o _udivdi3_s.o _umoddi3_s.o _udiv_w_sdiv_s.o _udivmoddi4_s.o cpuinfo_s.o tf-signs_s.o sfp-exceptions_s.o addtf3_s.o divtf3_s.o eqtf2_s.o getf2_s.o letf2_s.o multf3_s.o negtf2_s.o subtf3_s.o unordtf2_s.o fixtfsi_s.o fixunstfsi_s.o floatsitf_s.o floatunsitf_s.o fixtfdi_s.o fixunstfdi_s.o floatditf_s.o floatunditf_s.o extendsftf2_s.o extenddftf2_s.o extendxftf2_s.o trunctfsf2_s.o trunctfdf2_s.o trunctfxf2_s.o enable-execute-stack_s.o unwind-dw2_s.o unwind-dw2-fde_s.o unwind-sjlj_s.o unwind-c_s.o emutls_s.o libgcc.a -Wl,-lpthread -lmingwthrd -lmingw32 -lmingwex -lmoldname -lmsvcrt -ladvapi32 -lshell32 -luser32 -lkernel32 && if [ -f ./shlib/libgcc_s_dw2-1.dll ]; then mv -f ./shlib/libgcc_s_dw2-1.dll ./shlib/libgcc_s_dw2-1.dll.backup; else true; fi && mv ./shlib/libgcc_s_dw2-1.dll.tmp ./shlib/libgcc_s_dw2-1.dll && mv ./shlib/libgcc_s.a.tmp ./shlib/libgcc_s.a
__main_s.o: In function `_do_global_dtors':
/Users/tommy/mxe/tmp-gcc-i686-w64-mingw32.shared/gcc-5.1.0/libgcc/libgcc2.c:2240: undefined reference to `__EH_FRAME_BEGIN__'
__main_s.o: In function `_do_global_ctors':
/Users/tommy/mxe/tmp-gcc-i686-w64-mingw32.shared/gcc-5.1.0/libgcc/libgcc2.c:2256: undefined reference to `__EH_FRAME_BEGIN__'
collect2: error: ld returned 1 exit status
make[4]: *** [libgcc_s.dll] Error 1
make[3]: *** [all-target-libgcc] Error 2
make[2]: *** [all] Error 2
make[1]: *** [build-only-gcc_i686-w64-mingw32.shared] Error 2

real	14m24.032s
user	28m12.427s
sys	6m12.026s

**Comment by aquiles2k** [comment]
There's some discussion here:
http://sourceforge.net/p/mingw-w64/mailman/message/34101954/

And a partly patch here:
https://gcc.gnu.org/ml/gcc-patches/2015-05/msg01212.html

**Comment by mabreedlove** [comment]
Saw this while writing up the bug.  Patch that restores pre-5.0 behavior is attached to the bug.

https://gcc.gnu.org/bugzilla/show_bug.cgi?id=67172

**Comment by tonytheodore** [comment]
Still happening in 5.2 with `--disable-sjlj-exceptions`

> I'm building for c++11 threads using the two stage compilation explained here: http://stackoverflow.com/a/24538951

See #958, you'll be able to build posix threads with either:
```
make gcc MXE_TARGETS=i686-w64-mingw32.shared MXE_GCC_THREADS=posix
```
or
```
make gcc MXE_TARGETS=i686-w64-mingw32.shared.posix
```

**Comment by aquiles2k** [comment]
Ok, cool, so if i understand correctly i can build posix threads now in one step with "--with-dwarf2" but not yet with "--disable-sjlj-exceptions" ?

**Comment by tonytheodore** [comment]
Not exactly, threads and exceptions are orthogonal. You'll be able to build posix threads in one step, exceptions have always been one step but I'm no longer certain of the defaults.

I thought we were using `dw2` on i686 and `seh` on x86_64, but [`gdb` made me question that](http://lists.nongnu.org/archive/html/mingw-cross-env-list/2015-10/msg00035.html). We're going to have to set these explicitly (and enable options) but testing the combinations is difficult.

P.S. `--with-dwarf2` relates to debugging symbols - not exception handling.

**Comment by tonytheodore** [comment]
Actually, since we don't set the options, you could trick it with:
```
make gcc MXE_TARGETS=i686-w64-mingw32.shared MXE_GCC_THREADS='posix --disable-sjlj-exceptions'
```
There will be better ways in the future.

**Comment by aquiles2k** [comment]
Ok i successfuly build posix gcc in one step and with both flags, with your changes and the patch by Matt Breedlove in the gcc bugreport!

Unfortunately, since i needed this for qt but now found out webkit is disabled in qt5.5 i still have to use qt5.4 :( Has anybody tried to build it nevertheless? Simply reactivating didnt seem to do the trick.

Apart from that this issue might be closed.

**Comment by tonytheodore** [comment]
Note gcc is about to be downgraded to 4.9.3 in #965 and it may (or may not) have something to do with qtwebkit build failures (#747).

**Comment by aquiles2k** [comment]
Ok, ill watch and retry it when this happens.

**Comment by aquiles2k** [comment]
Unfortunately after recent gcc downgrade webkit it still fails with the same error ([qtwebkit_i686-w64-mingw32.shared.txt](https://github.com/mxe/mxe/files/36723/qtwebkit_i686-w64-mingw32.shared.txt)
):

```
➜  mxe git:(master) make qtwebkit MXE_TARGETS=i686-w64-mingw32.shared MXE_GCC_THREADS='posix --disable-sjlj-exceptions --with-dwarf2'
[build]     qtwebkit               i686-w64-mingw32.shared

Failed to build package qtwebkit for target i686-w64-mingw32.shared!
------------------------------------------------------------
/bin/sh: line 0: type: nul: not found
make[4]: *** [generated/supplemental_dependency.tmp] Error 1
make[4]: *** Waiting for unfinished jobs....
make[3]: *** [sub-DerivedSources-pri-make_first-ordered] Error 2
make[2]: *** [sub-Source-WebCore-WebCore-pro-make_first-ordered] Error 2
make[1]: *** [build-only-qtwebkit_i686-w64-mingw32.shared] Error 2
real	7m4.605s
user	19m48.903s
sys	1m39.295s
------------------------------------------------------------
[log]      /Users/aj/dev/mxe/log/qtwebkit_i686-w64-mingw32.shared

make: *** [/Users/aj/dev/mxe/usr/i686-w64-mingw32.shared/installed/qtwebkit] Error 1
```

Should i open a new issue, append to #747 or leave it be since webkit is deprecated in qt5.5 (even though precompiled win binaries still ship with it - 5.6 ist announced to have it removed)?

**Comment by mabrand** [comment]
Doesn't qtwebkit in Qt 5.5.1 require Angle support? As far as I know MXE's winapi doesn't include Angle.

**Comment by aquiles2k** [comment]
Im not sure about that. I tried to find a statement but couldnt find something explicit.

Theres for example the pretty old http://trac.webkit.org/wiki/BuildingQtOnWindows which doesnt state anything in regards to Angle or DirectX.

Theres https://wiki.qt.io/Building_Qt_5_from_Git and http://doc.qt.io/qt-5/windows-building.html where i couldnt find anything.

All just state various tools like perl, python, ruby, flex and the icu lib.

Do you have some other sources stating that?

Besides Digia themselves releasing binary packages with qt 5.5.1 webkit theres also msys2 which has webkit packages. You can even find various patches in their package repo in that regard.

See e.g. https://github.com/Alexpux/MINGW-packages/blob/master/mingw-w64-qt5/0021-qt-5.3.0-qtwebkit-dont-use-bundled-angle-libraries.patch

**Comment by aquiles2k** [comment]
I applied the patch from here: https://github.com/Alexpux/Qt-builds/blob/master/patches/qt/5.0.x/qt-5.0.0-fix-build-under-msys.patch and just saw that the build completed - Yay :)

Have yet to test it with our app if theres anything buggy.

**Comment by tonytheodore** [comment]
That patch seems to work so I've [re-enabled the build](https://github.com/mxe/mxe/commit/19d0acd474e39fbc2c2f7f136f9abaafdfcdc337).

**Comment by aquiles2k** [comment]
Yes, our App is live now for a week with 5.5.1 webkit without problems. Thx all

**Comment by aquiles2k** [comment]
Just as heads up regarding webkit with upcoming 5.6 release. Comment from Thiago Macieira (December 19, 2015 at 05:39) in 5.6 Beta Anouncement: "Actually, QtWebKit is guaranteed to work for at least a couple more versions. We’ve just revisited the decision and will release a version 5.6 of it that does compile. But it will not be included in the main binaries. You’ll need to build from sources."

**Comment by tonytheodore** [comment]
Thanks, that's good to know. It also appears that 5.6 will be supported for three years so there will be plenty of time to transition.

**Comment by rr-** [comment]
How's progress on simplifying the `MXE_GCC_THREADS=posix` switches?

**Comment by tonytheodore** [comment]
Not much progress lately, on holidays now so should be able to look at it again soon.

Note the recommended invocation will be:
```
make gcc MXE_TARGETS=i686-w64-mingw32.shared.posix
```

**Comment by rr-** [comment]
Thank you for the answer @tonytheodore - just one more question though, are static builds discouraged for any reason?

**Comment by tonytheodore** [comment]
Static builds are not discouraged, if anything I'd recommend them in almost all cases. I meant the general form of future targets will be:

```
<triplet>.<link>.<threads>.<exceptions>

i686-w64-mingw32.static.posix.dw2
x86_64-w64-mingw32.shared.win32.sjlj
```

**Comment by starius** [comment]
Will `i686-w64-mingw32.static` work?

**Comment by tonytheodore** [comment]
Yes, the current forms will continue to work and use the current settings:
 * `win32` threads
 * `sjlj` for `i686` and `seh` for `x86_64`

The additional segments are only required to use the non-default options.

**Comment by aquiles2k** [comment]
Just as a heads up regarding the original issue here with __EH_FRAME_BEGIN__ being undefined in gcc 5.x with --disable-sjlj-exceptions and --with-dwarf2 which is documented in https://gcc.gnu.org/bugzilla/show_bug.cgi?id=67172

According to the bugreport it seems to be fixed now in gcc 5.4
